### PR TITLE
17289 enforce minimum password strength

### PR DESCRIPTION
**Pull request by arthanson**
### Fixes: #17289 

Enforce default minimum password strength:
* 12 characters minimum
* similarities to user attributes (first name, last name, email)
* common password check
* must contain one number, one upper-case letter and one lower-case letter.

Change Password | NetBox 2024-08-29 08-58-46

### CODE REVIEW

File: netbox/users/tests/test_views.py [hunk 1]
```diff
@@ -60,19 +60,32 @@ def setUpTestData(cls):
             'last_name': 'newlastname',
         }
 
-    @override_settings(AUTH_PASSWORD_VALIDATORS=[{
-        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
-        'OPTIONS': {'min_length': 8}
-    }])
+    @override_settings(AUTH_PASSWORD_VALIDATORS=[
```
Comment (jeremystretch) [inline netbox/users/tests/test_views.py:63]: Why not just leave the default policy in place?

File: netbox/users/tests/test_api.py
Comment (jeremystretch) [inline netbox/users/tests/test_api.py:1]: Same notes here as for `test_views.py`.

File: netbox/utilities/password_validation.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext as _
+
+
+class NumericAlphaPasswordValidator:
```
Comment (jeremystretch) [inline netbox/utilities/password_validation.py:5]:
```suggestion
class AlphanumericPasswordValidator:
```

File: netbox/utilities/password_validation.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    """
+    Validate that the password has at least one numeral, one uppercase letter and one lowercase letter.
+    """
+
+    def validate(self, password, user=None):
+        if not any(char.isdigit() for char in password):
+            raise ValidationError(
+                _("Password should have at least one numeral"),
```
Comment (jeremystretch) [inline netbox/utilities/password_validation.py:13]:
```suggestion
                _("Password must have at least one numeral."),
```

File: netbox/utilities/password_validation.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            )
+
+        if not any(char.isupper() for char in password):
+            raise ValidationError(
+                _("Password should have at least one uppercase letter"),
```
Comment (jeremystretch) [inline netbox/utilities/password_validation.py:18]:
```suggestion
                _("Password must have at least one uppercase letter."),
```

File: netbox/utilities/password_validation.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            )
+
+        if not any(char.islower() for char in password):
+            raise ValidationError(
+                _("Password should have at least one lowercase letter"),
```
Comment (jeremystretch) [inline netbox/utilities/password_validation.py:23]:
```suggestion
                _("Password must have at least one lowercase letter."),
```

File: netbox/users/tests/test_views.py [hunk 6]
```diff
@@ -84,10 +97,30 @@ def test_password_validation_enforced(self):
...
+
+        # Password no uppercase
+        data['password'] = 'foobar123abc'
+        data['confirm_password'] = 'foobar123abc'
+        self.assertHttpStatus(self.client.post(**request), 200)
+
+        # Password no lowercase
+        data['password'] = 'FOOBAR123ABC'
+        data['confirm_password'] = 'FOOBAR123ABC'
+        self.assertHttpStatus(self.client.post(**request), 200)
```
Comment (jeremystretch) [inline netbox/users/tests/test_views.py:122]: We can move the individual validation logic tests to a separate test case user `users`; there's no need to test the validators for both UI and API requests.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/netbox/settings.py [hunk 7]
```diff
@@ -63,7 +62,23 @@
 ADMINS = getattr(configuration, 'ADMINS', [])
 ALLOW_TOKEN_RETRIEVAL = getattr(configuration, 'ALLOW_TOKEN_RETRIEVAL', True)
 ALLOWED_HOSTS = getattr(configuration, 'ALLOWED_HOSTS')  # Required
-AUTH_PASSWORD_VALIDATORS = getattr(configuration, 'AUTH_PASSWORD_VALIDATORS', [])
+AUTH_PASSWORD_VALIDATORS = getattr(configuration, 'AUTH_PASSWORD_VALIDATORS', [
+    {
+        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
```
Comment (jeremystretch) [inline netbox/netbox/settings.py:67]: This doesn't seem to validate when creating a new user. For example, I can create a user with the name & password of `NewUser123456`. However, if I then edit the user and set the same password, I get a validation error complaining that the password is too similar to the username.

AFAICT this is all stock Django functionality so I'm not sure what's going on here.

File: netbox/netbox/settings.py [hunk 8]
```diff
@@ -63,7 +62,23 @@
...
+    },
+    {
+        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
+        "OPTIONS": {
+            "min_length": 12,
+        },
+    },
+    {
+        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
```
Comment (jeremystretch) [inline netbox/netbox/settings.py:76]: IMO we should omit this validator, for two reasons:

1. We don't control the list of "common" passwords, and it doesn't appear to be formally documented.
2. Our minimum length validation will disqualify nearly all of [these strings](https://gist.github.com/roycewilliams/226886fd01572964e1431ac8afc999ce) anyway.

File: netbox/users/tests/test_api.py [hunk 9]
```diff
@@ -111,10 +111,30 @@ def test_password_validation_enforced(self):
...
+
+        # Password no uppercase
+        data['password'] = 'foobarfoo1'
+        response = self.client.post(url, data, format='json', **self.header)
+        self.assertEqual(response.status_code, 400)
+
+        # Password no lowercase
+        data['password'] = 'FOOBARFOO1'
+        response = self.client.post(url, data, format='json', **self.header)
+        self.assertEqual(response.status_code, 400)
```
Comment (jeremystretch) [inline netbox/users/tests/test_api.py:136]: This set of logic tests is duplicated for the UI and API. Can we move it to a single test module dedicated to testing password validation on the User model itself?

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/users/tests/test_api.py (hunk 9 above)
Comment (arthanson) [reply netbox/users/tests/test_api.py:136]: IMHO: I think it needs to be here as https://github.com/netbox-community/netbox/blob/develop/netbox/users/forms/model_forms.py#L231 specifically calls validate_password so these checks make sure it is getting called appropriately in the form.

File: netbox/netbox/settings.py (hunk 7 above)
Comment (arthanson) [reply netbox/netbox/settings.py:67]: that was tricky, needs to be in _post_clean as in clean self.instance isn't set on create.

**Review by jeremystretch — APPROVED** [review]

**Pull request merged**